packages/editor/src/components/document-outline/index.js

```diff
@@ -113,104 +148,145 @@ export default function DocumentOutline( {
 	hasOutlineItemsDisabled,
 } ) {
 	const { selectBlock } = useDispatch( blockEditorStore );
-	const { blocks, title, isTitleSupported } = useSelect( ( select ) => {
-		const { getBlocks } = select( blockEditorStore );
-		const { getEditedPostAttribute } = select( editorStore );
-		const { getPostType } = select( coreStore );
-		const postType = getPostType( getEditedPostAttribute( 'type' ) );
-
-		return {
-			title: getEditedPostAttribute( 'title' ),
-			blocks: getBlocks(),
-			isTitleSupported: postType?.supports?.title ?? false,
-		};
-	} );
+	const { blocks, title, isTitleSupported, isShowingTemplate } = useSelect(
+		( select ) => {
+			const { getBlocks } = select( blockEditorStore );
+			const { getEditedPostAttribute, getRenderingMode } =
+				select( editorStore );
+			const { getPostType } = select( coreStore );
+			const postType = getPostType( getEditedPostAttribute( 'type' ) );
 
+			return {
+				title: getEditedPostAttribute( 'title' ),
+				blocks: getBlocks(),
+				isTitleSupported: postType?.supports?.title ?? false,
+				isShowingTemplate:
+					postType?.slug === 'wp_template' ||
+					getRenderingMode() === 'template-locked',
+			};
+		}
+	);
 	const prevHeadingLevelRef = useRef( 1 );
+	const outlineElements = computeOutlineElements( blocks, isShowingTemplate );
+	const headings = outlineElements.filter(
+		( item ) => item.type === 'heading'
+	);
+	const mainElements = outlineElements.filter(
+		( item ) => item.type === 'main'
+	);
+	const headingsByLevel = headings.reduce( ( acc, heading ) => {
+		acc[ heading.level ] = ( acc[ heading.level ] || 0 ) + 1;
+		return acc;
+	}, {} );
+	const hasMultipleH1 = headingsByLevel[ 1 ] > 1;
+
+	/**
+	 * TODO: Update hasTitle to work with the iframe in the block editor.
+	 * The titleNode is not found with the querySelector when the editor is iframed.
+	 * See https://github.com/WordPress/gutenberg/issues/47624
+	 */
+	const titleNode = document.querySelector( '.editor-post-title__input' );
+	const hasTitle = isTitleSupported && title && titleNode;
 
-	const headings = computeOutlineHeadings( blocks );
-	if ( headings.length < 1 ) {
-		return (
-			<div className="editor-document-outline has-no-headings">
-				<EmptyOutlineIllustration />
+	return (
+		<div
+			className={ clsx(
+				'document-outline',
+				headings.length < 1 && 'has-no-headings',
+				isShowingTemplate && mainElements.length === 0 && 'has-no-main'
+			) }
+		>
+			{ headings.length < 1 && (
+				<>
+					<EmptyOutlineIllustration />
+					<p>
+						{ __(
+							'Navigate the structure of your document and address issues like empty or incorrect heading levels.'
+						) }
+					</p>
+				</>
+			) }
+			{ isShowingTemplate && mainElements.length === 0 && (
 				<p>
```

Comment:
I'd suggest to wrap this warning in a `Notice` component, which is included in `@wordpress/components`. Here's the source code I've tested it with locally:

```
<Notice
spokenMessage={ null }
status="warning"
isDismissible={ false }
>
<p>
	{ __(
		'The main element is missing. Select the block that contains your most important content and add the main HTML element in the Advanced panel.'
	) }
</p>
</Notice>
```